design/20220118.server-side-apply.md

```diff
@@ -0,0 +1,242 @@
+one controller (issuing and trigger Certificate controllers, and cmctl), and as
+such, will need to make use of the `force` parameter in their API calls. This
+option tells the API server to revoke management of that field from the previous
+owner, overwrite the field, and change owner ship to the new client. Since some
+fields, such as the Issuing Condition, may have an undefined number of potential
+managers (both internal and external to the cert-manager controller), using the
+same manager for things is not a possibility. You can read more about the
+`force` paramerter on the Kubernetes documentation on
+[Server Side Apply](https://kubernetes.io/docs/reference/using-api/server-side-apply/),
+and in particular the
+[Conflicts](https://kubernetes.io/docs/reference/using-api/server-side-apply/#conflicts)
+section.
+
+Use of the `force` parameter will not overwrite other controllers progress
+since, 1 controllers only update the fields that they manage, 2 controllers who
+manage the same fields should never be writing to the same field at the same
+time (as they should be gated by those or other fields presence and values, such
+as the Issuing condition for the issuing and trigger controllers).
+
+There is likely no APPLY call that cert-manager does not set the `force`
+parameter to true since it, [never wants to give up ownership claim, and always
+wants to overwrite
+values](https://kubernetes.io/docs/reference/using-api/server-side-apply/#conflicts).
+See
+[here](https://kubernetes.io/docs/reference/using-api/server-side-apply/#using-server-side-apply-in-a-controller).
+
+### client-go Testing
+
+The [fake client-go client](https://github.com/kubernetes/client-go/issues/970)
+does not support the Apply patch call for mocking API calls and events. This
+means that significant controller unit-testing will either need to moved to
+testing against a real API server as integration tests, the controller
+[test framework must add custom support for Apply](https://github.com/jetstack/cert-manager/blob/master/pkg/controller/test/context_builder.go),
+or a new testing framework should be developed. We can also PR this upstream but
+will take time to be released so a stop gap would always be needed.
+
+### API Changes
+
+Some API fields will need to have some metadata updated to function better under
+Server Side Apply. One such example is adding `x-kubernetes-list-type=map` and
+`x-kubernetes-list-map-keys=Type` to the [Certificates Status Condition
+slice](https://github.com/jetstack/cert-manager/blob/0ca1ce9a6a1d7c311afd4b3e786975759249132a/pkg/apis/certmanager/v1/types_certificate.go#L385),
+so that controllers are able to apply distinct condition types, without
+conflicting with other controller conditions (i.e. the Ready and Issuing
+conditions). Integration tests will be able to ensure cert-manager have set
+these API type tags correctly.
+
+## Feature Gate
+
+Placing the APPLY functionality behind a feature gate should be required.
```

Comment:
Is there a reason for using the typography `APPLY`? I say that because the all-caps makes me think it is an HTTP verb, but it seems "Apply" is actually a `PATCH` with the `Content-Type: application/apply-patch+yaml` ([patch.go](https://github.com/kubernetes/apimachinery/blob/57f2a0733447cfd41294477d833cce6580faaca3/pkg/types/patch.go#L28)), as per [555-server-side-apply.md](https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/555-server-side-apply/README.md#proposal):

> Apply is invoked by sending a certain `Content-Type` with the verb PATCH.

Reply:
Fixed in https://github.com/jetstack/cert-manager/pull/4758/commits/bfcaa3dd55c7f7c8e98c9775dfd12fbb49d514aa